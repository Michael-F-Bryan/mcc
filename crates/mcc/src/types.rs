use std::{
    fmt::{self, Display},
    ops::Deref,
};

use mcc_syntax::ast::TranslationUnit;
use type_sitter::Node;

use crate::{Db, Text};

#[derive(mcc_macros::SerializeWithDatabase)]
#[salsa::input]
#[derive(Debug)]
pub struct SourceFile {
    #[returns(ref)]
    pub path: Text,
    #[returns(ref)]
    pub contents: Text,
}

#[salsa::tracked]
pub struct Ast<'db> {
    #[returns(ref)]
    pub tree: Tree,
}

#[salsa::tracked]
impl<'db> Ast<'db> {
    pub fn sexpr(&self, db: &'db dyn Db) -> impl Display {
        SexpPrinter::new(self.tree(db).root_node())
    }

    pub fn root(&self, db: &'db dyn Db) -> TranslationUnit<'db> {
        let root = self.tree(db).root_node();
        TranslationUnit::try_from_raw(root).unwrap()
    }
}

#[derive(Clone)]
struct SexpPrinter<'db> {
    node: tree_sitter::Node<'db>,
    indent: usize,
}

impl<'db> SexpPrinter<'db> {
    fn new(node: tree_sitter::Node<'db>) -> Self {
        Self { node, indent: 0 }
    }
}

impl<'db> Display for SexpPrinter<'db> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fn write_indent(f: &mut fmt::Formatter<'_>, indent: usize) -> fmt::Result {
            for _ in 0..indent {
                write!(f, "  ")?;
            }
            Ok(())
        }

        fn escape_quoted_atom(s: &str) -> String {
            let mut out = String::with_capacity(s.len() + 4);
            for ch in s.chars() {
                match ch {
                    '"' => out.push_str("\\\""),
                    '\\' => out.push_str("\\\\"),
                    _ => out.push(ch),
                }
            }
            out
        }

        fn write_node(
            f: &mut fmt::Formatter<'_>,
            node: tree_sitter::Node<'_>,
            indent: usize,
            emit_indent: bool,
        ) -> fmt::Result {
            if emit_indent {
                write_indent(f, indent)?;
            }

            // Handle missing nodes specially to match tree-sitter's S-expression output
            if node.is_missing() {
                let kind = node.kind();
                let quoted = escape_quoted_atom(kind);
                write!(f, "(MISSING \"{quoted}\")")?;
                return Ok(());
            }

            let is_error = node.is_error();
            if is_error {
                write!(f, "(ERROR")?;
            } else {
                let k = node.kind();
                if k.contains(|c: char| !c.is_alphanumeric() && c != '_') {
                    write!(f, "(\"{k}\"")?;
                } else {
                    write!(f, "({k}")?;
                }
            }

            let child_count = node.child_count();
            if child_count > 0 {
                // Each child on its own line, indented one level deeper
                for i in 0..child_count {
                    // Safety: index is in-bounds per child_count
                    if let Some(child) = node.child(i) {
                        writeln!(f)?;
                        write_indent(f, indent + 1)?;

                        if let Some(field_name) = node.field_name_for_child(i as u32) {
                            write!(f, "{field_name}: ")?;
                            // Child already has indentation emitted above; don't emit again here
                            write_node(f, child, indent + 1, false)?;
                        } else {
                            write_node(f, child, indent + 1, false)?;
                        }
                    }
                }
            }

            write!(f, ")")
        }

        write_node(f, self.node, self.indent, true)
    }
}

#[derive(Debug, Clone)]
pub struct Tree(pub tree_sitter::Tree);

impl From<tree_sitter::Tree> for Tree {
    fn from(value: tree_sitter::Tree) -> Self {
        Tree(value)
    }
}

impl Deref for Tree {
    type Target = tree_sitter::Tree;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl PartialEq for Tree {
    fn eq(&self, other: &Self) -> bool {
        self.0.root_node() == other.0.root_node()
    }
}

impl Eq for Tree {}

impl std::hash::Hash for Tree {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.0.root_node().hash(state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! sexpr_test {
        (
            $(
                $(#[$meta:meta])*
                $name:ident : $sexpr:expr
            ),*
            $(,)?
        ) => {
            $(
                $(#[$meta])*
                #[test]
                fn $name() {
                    let mut parser = tree_sitter::Parser::new();
                    parser.set_language(&tree_sitter_c::LANGUAGE.into()).unwrap();
                    let tree = parser.parse($sexpr, None).unwrap();
                    let root = tree.root_node();

                    println!("{}", root.to_sexp());
                    insta::with_settings!(
                        { description => $sexpr },
                        {
                            let formatted = SexpPrinter::new(root).to_string();
                            insta::assert_snapshot!(formatted);
                        }
                    );
                }
            )*
        };
    }

    sexpr_test! {
        empty: "",
        comment: "/* */",
        declaration: "int x;",
        function_definition: "void main();",
        function_call: "void main(void)",
        unclosed_paren: "int (",
    }
}
