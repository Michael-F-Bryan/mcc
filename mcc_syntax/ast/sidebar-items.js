window.SIDEBAR_ITEMS = {"enum":["AbstractDeclarator","Declarator","Expression","FieldDeclarator","Statement","TypeDeclarator","TypeSpecifier"],"mod":["anon_unions","symbols","unnamed"],"struct":["AbstractArrayDeclarator","AbstractFunctionDeclarator","AbstractParenthesizedDeclarator","AbstractPointerDeclarator","AlignasQualifier","AlignofExpression","ArgumentList","ArrayDeclarator","AssignmentExpression","Attribute","AttributeDeclaration","AttributeSpecifier","AttributedDeclarator","AttributedStatement","BinaryExpression","BitfieldClause","BreakStatement","CallExpression","CaseStatement","CastExpression","CharLiteral","Character","CommaExpression","Comment","CompoundLiteralExpression","CompoundStatement","ConcatenatedString","ConditionalExpression","ContinueStatement","Declaration","DeclarationList","DoStatement","ElseClause","EnumSpecifier","Enumerator","EnumeratorList","EscapeSequence","ExpressionStatement","ExtensionExpression","False","FieldDeclaration","FieldDeclarationList","FieldDesignator","FieldExpression","FieldIdentifier","ForStatement","FunctionDeclarator","FunctionDefinition","GenericExpression","GnuAsmClobberList","GnuAsmExpression","GnuAsmGotoList","GnuAsmInputOperand","GnuAsmInputOperandList","GnuAsmOutputOperand","GnuAsmOutputOperandList","GnuAsmQualifier","GotoStatement","Identifier","IfStatement","InitDeclarator","InitializerList","InitializerPair","LabeledStatement","LinkageSpecification","MacroTypeSpecifier","MsBasedModifier","MsCallModifier","MsDeclspecModifier","MsPointerModifier","MsRestrictModifier","MsSignedPtrModifier","MsUnalignedPtrModifier","MsUnsignedPtrModifier","Null","NumberLiteral","OffsetofExpression","ParameterDeclaration","ParameterList","ParenthesizedDeclarator","ParenthesizedExpression","PointerDeclarator","PointerExpression","PreprocArg","PreprocCall","PreprocDef","PreprocDefined","PreprocDirective","PreprocElif","PreprocElifdef","PreprocElse","PreprocFunctionDef","PreprocIf","PreprocIfdef","PreprocInclude","PreprocParams","PrimitiveType","ReturnStatement","SehExceptClause","SehFinallyClause","SehLeaveStatement","SehTryStatement","SizedTypeSpecifier","SizeofExpression","StatementIdentifier","StorageClassSpecifier","StringContent","StringLiteral","StructSpecifier","SubscriptDesignator","SubscriptExpression","SubscriptRangeDesignator","SwitchStatement","SystemLibString","TranslationUnit","True","TypeDefinition","TypeDescriptor","TypeIdentifier","TypeQualifier","UnaryExpression","UnionSpecifier","UpdateExpression","VariadicParameter","WhileStatement"]};