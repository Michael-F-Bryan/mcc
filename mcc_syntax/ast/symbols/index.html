<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `symbols` mod in crate `mcc_syntax`."><title>mcc_syntax::ast::symbols - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-1a91846b.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="mcc_syntax" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0 (6b00bc388 2025-06-23)" data-channel="1.88.0" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../mcc_syntax/index.html">mcc_<wbr>syntax</a><span class="version">0.0.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module symbols</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In mcc_<wbr>syntax::<wbr>ast</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">mcc_syntax</a>::<wbr><a href="../index.html">ast</a></div><h1>Module <span>symbols</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/mcc_syntax/home/runner/work/mcc/mcc/target/debug/build/mcc-syntax-ba4364fc9a529433/out/nodes.rs.html#14456">Source</a> </span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.Add.html" title="struct mcc_syntax::ast::symbols::Add">Add</a></dt><dd>Typed node <code>+</code></dd><dt><a class="struct" href="struct.AddAdd.html" title="struct mcc_syntax::ast::symbols::AddAdd">AddAdd</a></dt><dd>Typed node <code>++</code></dd><dt><a class="struct" href="struct.AddEq.html" title="struct mcc_syntax::ast::symbols::AddEq">AddEq</a></dt><dd>Typed node <code>+=</code></dd><dt><a class="struct" href="struct.And.html" title="struct mcc_syntax::ast::symbols::And">And</a></dt><dd>Typed node <code>&amp;</code></dd><dt><a class="struct" href="struct.AndAnd.html" title="struct mcc_syntax::ast::symbols::AndAnd">AndAnd</a></dt><dd>Typed node <code>&amp;&amp;</code></dd><dt><a class="struct" href="struct.AndEq.html" title="struct mcc_syntax::ast::symbols::AndEq">AndEq</a></dt><dd>Typed node <code>&amp;=</code></dd><dt><a class="struct" href="struct.BitNot.html" title="struct mcc_syntax::ast::symbols::BitNot">BitNot</a></dt><dd>Typed node <code>~</code></dd><dt><a class="struct" href="struct.BitXor.html" title="struct mcc_syntax::ast::symbols::BitXor">BitXor</a></dt><dd>Typed node <code>^</code></dd><dt><a class="struct" href="struct.BitXorEq.html" title="struct mcc_syntax::ast::symbols::BitXorEq">BitXor<wbr>Eq</a></dt><dd>Typed node <code>^=</code></dd><dt><a class="struct" href="struct.Colon.html" title="struct mcc_syntax::ast::symbols::Colon">Colon</a></dt><dd>Typed node <code>:</code></dd><dt><a class="struct" href="struct.ColonColon.html" title="struct mcc_syntax::ast::symbols::ColonColon">Colon<wbr>Colon</a></dt><dd>Typed node <code>::</code></dd><dt><a class="struct" href="struct.Comma.html" title="struct mcc_syntax::ast::symbols::Comma">Comma</a></dt><dd>Typed node <code>,</code></dd><dt><a class="struct" href="struct.Div.html" title="struct mcc_syntax::ast::symbols::Div">Div</a></dt><dd>Typed node <code>/</code></dd><dt><a class="struct" href="struct.DivEq.html" title="struct mcc_syntax::ast::symbols::DivEq">DivEq</a></dt><dd>Typed node <code>/=</code></dd><dt><a class="struct" href="struct.Dot.html" title="struct mcc_syntax::ast::symbols::Dot">Dot</a></dt><dd>Typed node <code>.</code></dd><dt><a class="struct" href="struct.DotDotDot.html" title="struct mcc_syntax::ast::symbols::DotDotDot">DotDot<wbr>Dot</a></dt><dd>Typed node <code>...</code></dd><dt><a class="struct" href="struct.DoubleQuote.html" title="struct mcc_syntax::ast::symbols::DoubleQuote">Double<wbr>Quote</a></dt><dd>Typed node <code>"</code></dd><dt><a class="struct" href="struct.Eq.html" title="struct mcc_syntax::ast::symbols::Eq">Eq</a></dt><dd>Typed node <code>=</code></dd><dt><a class="struct" href="struct.EqEq.html" title="struct mcc_syntax::ast::symbols::EqEq">EqEq</a></dt><dd>Typed node <code>==</code></dd><dt><a class="struct" href="struct.Gt.html" title="struct mcc_syntax::ast::symbols::Gt">Gt</a></dt><dd>Typed node <code>&gt;</code></dd><dt><a class="struct" href="struct.GtEq.html" title="struct mcc_syntax::ast::symbols::GtEq">GtEq</a></dt><dd>Typed node <code>&gt;=</code></dd><dt><a class="struct" href="struct.GtGt.html" title="struct mcc_syntax::ast::symbols::GtGt">GtGt</a></dt><dd>Typed node <code>&gt;&gt;</code></dd><dt><a class="struct" href="struct.GtGtEq.html" title="struct mcc_syntax::ast::symbols::GtGtEq">GtGtEq</a></dt><dd>Typed node <code>&gt;&gt;=</code></dd><dt><a class="struct" href="struct.Hashdefine.html" title="struct mcc_syntax::ast::symbols::Hashdefine">Hashdefine</a></dt><dd>Typed node <code>#define</code></dd><dt><a class="struct" href="struct.Hashelif.html" title="struct mcc_syntax::ast::symbols::Hashelif">Hashelif</a></dt><dd>Typed node <code>#elif</code></dd><dt><a class="struct" href="struct.Hashelifdef.html" title="struct mcc_syntax::ast::symbols::Hashelifdef">Hashelifdef</a></dt><dd>Typed node <code>#elifdef</code></dd><dt><a class="struct" href="struct.Hashelifndef.html" title="struct mcc_syntax::ast::symbols::Hashelifndef">Hashelifndef</a></dt><dd>Typed node <code>#elifndef</code></dd><dt><a class="struct" href="struct.Hashelse.html" title="struct mcc_syntax::ast::symbols::Hashelse">Hashelse</a></dt><dd>Typed node <code>#else</code></dd><dt><a class="struct" href="struct.Hashendif.html" title="struct mcc_syntax::ast::symbols::Hashendif">Hashendif</a></dt><dd>Typed node <code>#endif</code></dd><dt><a class="struct" href="struct.Hashif.html" title="struct mcc_syntax::ast::symbols::Hashif">Hashif</a></dt><dd>Typed node <code>#if</code></dd><dt><a class="struct" href="struct.Hashifdef.html" title="struct mcc_syntax::ast::symbols::Hashifdef">Hashifdef</a></dt><dd>Typed node <code>#ifdef</code></dd><dt><a class="struct" href="struct.Hashifndef.html" title="struct mcc_syntax::ast::symbols::Hashifndef">Hashifndef</a></dt><dd>Typed node <code>#ifndef</code></dd><dt><a class="struct" href="struct.Hashinclude.html" title="struct mcc_syntax::ast::symbols::Hashinclude">Hashinclude</a></dt><dd>Typed node <code>#include</code></dd><dt><a class="struct" href="struct.LBrace.html" title="struct mcc_syntax::ast::symbols::LBrace">LBrace</a></dt><dd>Typed node <code>{</code></dd><dt><a class="struct" href="struct.LBracket.html" title="struct mcc_syntax::ast::symbols::LBracket">LBracket</a></dt><dd>Typed node <code>[</code></dd><dt><a class="struct" href="struct.LBracketLBracket.html" title="struct mcc_syntax::ast::symbols::LBracketLBracket">LBracketL<wbr>Bracket</a></dt><dd>Typed node <code>[[</code></dd><dt><a class="struct" href="struct.LDoubleQuote.html" title="struct mcc_syntax::ast::symbols::LDoubleQuote">LDouble<wbr>Quote</a></dt><dd>Typed node <code>L"</code></dd><dt><a class="struct" href="struct.LParen.html" title="struct mcc_syntax::ast::symbols::LParen">LParen</a></dt><dd>Typed node <code>(</code></dd><dt><a class="struct" href="struct.LQuote.html" title="struct mcc_syntax::ast::symbols::LQuote">LQuote</a></dt><dd>Typed node <code>L'</code></dd><dt><a class="struct" href="struct.Lt.html" title="struct mcc_syntax::ast::symbols::Lt">Lt</a></dt><dd>Typed node <code>&lt;</code></dd><dt><a class="struct" href="struct.LtEq.html" title="struct mcc_syntax::ast::symbols::LtEq">LtEq</a></dt><dd>Typed node <code>&lt;=</code></dd><dt><a class="struct" href="struct.LtLt.html" title="struct mcc_syntax::ast::symbols::LtLt">LtLt</a></dt><dd>Typed node <code>&lt;&lt;</code></dd><dt><a class="struct" href="struct.LtLtEq.html" title="struct mcc_syntax::ast::symbols::LtLtEq">LtLtEq</a></dt><dd>Typed node <code>&lt;&lt;=</code></dd><dt><a class="struct" href="struct.Mod.html" title="struct mcc_syntax::ast::symbols::Mod">Mod</a></dt><dd>Typed node <code>%</code></dd><dt><a class="struct" href="struct.ModEq.html" title="struct mcc_syntax::ast::symbols::ModEq">ModEq</a></dt><dd>Typed node <code>%=</code></dd><dt><a class="struct" href="struct.Mul.html" title="struct mcc_syntax::ast::symbols::Mul">Mul</a></dt><dd>Typed node <code>*</code></dd><dt><a class="struct" href="struct.MulEq.html" title="struct mcc_syntax::ast::symbols::MulEq">MulEq</a></dt><dd>Typed node <code>*=</code></dd><dt><a class="struct" href="struct.Newline.html" title="struct mcc_syntax::ast::symbols::Newline">Newline</a></dt><dd>Typed node <code> </code></dd><dt><a class="struct" href="struct.Not.html" title="struct mcc_syntax::ast::symbols::Not">Not</a></dt><dd>Typed node <code>!</code></dd><dt><a class="struct" href="struct.NotEq.html" title="struct mcc_syntax::ast::symbols::NotEq">NotEq</a></dt><dd>Typed node <code>!=</code></dd><dt><a class="struct" href="struct.Or.html" title="struct mcc_syntax::ast::symbols::Or">Or</a></dt><dd>Typed node <code>|</code></dd><dt><a class="struct" href="struct.OrEq.html" title="struct mcc_syntax::ast::symbols::OrEq">OrEq</a></dt><dd>Typed node <code>|=</code></dd><dt><a class="struct" href="struct.OrOr.html" title="struct mcc_syntax::ast::symbols::OrOr">OrOr</a></dt><dd>Typed node <code>||</code></dd><dt><a class="struct" href="struct.Question.html" title="struct mcc_syntax::ast::symbols::Question">Question</a></dt><dd>Typed node <code>?</code></dd><dt><a class="struct" href="struct.Quote.html" title="struct mcc_syntax::ast::symbols::Quote">Quote</a></dt><dd>Typed node <code>'</code></dd><dt><a class="struct" href="struct.RBrace.html" title="struct mcc_syntax::ast::symbols::RBrace">RBrace</a></dt><dd>Typed node <code>}</code></dd><dt><a class="struct" href="struct.RBracket.html" title="struct mcc_syntax::ast::symbols::RBracket">RBracket</a></dt><dd>Typed node <code>]</code></dd><dt><a class="struct" href="struct.RBracketRBracket.html" title="struct mcc_syntax::ast::symbols::RBracketRBracket">RBracketR<wbr>Bracket</a></dt><dd>Typed node <code>]]</code></dd><dt><a class="struct" href="struct.RParen.html" title="struct mcc_syntax::ast::symbols::RParen">RParen</a></dt><dd>Typed node <code>)</code></dd><dt><a class="struct" href="struct.Semicolon.html" title="struct mcc_syntax::ast::symbols::Semicolon">Semicolon</a></dt><dd>Typed node <code>;</code></dd><dt><a class="struct" href="struct.Sub.html" title="struct mcc_syntax::ast::symbols::Sub">Sub</a></dt><dd>Typed node <code>-</code></dd><dt><a class="struct" href="struct.SubEq.html" title="struct mcc_syntax::ast::symbols::SubEq">SubEq</a></dt><dd>Typed node <code>-=</code></dd><dt><a class="struct" href="struct.SubGt.html" title="struct mcc_syntax::ast::symbols::SubGt">SubGt</a></dt><dd>Typed node <code>-&gt;</code></dd><dt><a class="struct" href="struct.SubSub.html" title="struct mcc_syntax::ast::symbols::SubSub">SubSub</a></dt><dd>Typed node <code>--</code></dd><dt><a class="struct" href="struct.U8DoubleQuote.html" title="struct mcc_syntax::ast::symbols::U8DoubleQuote">U8Double<wbr>Quote</a></dt><dd>Typed node <code>u8"</code></dd><dt><a class="struct" href="struct.U8Quote.html" title="struct mcc_syntax::ast::symbols::U8Quote">U8Quote</a></dt><dd>Typed node <code>u8'</code></dd><dt><a class="struct" href="struct.UDoubleQuote.html" title="struct mcc_syntax::ast::symbols::UDoubleQuote">UDouble<wbr>Quote</a></dt><dd>Typed node <code>U"</code></dd><dt><a class="struct" href="struct.UDoubleQuote_.html" title="struct mcc_syntax::ast::symbols::UDoubleQuote_">UDouble<wbr>Quote_</a></dt><dd>Typed node <code>u"</code></dd><dt><a class="struct" href="struct.UQuote.html" title="struct mcc_syntax::ast::symbols::UQuote">UQuote</a></dt><dd>Typed node <code>U'</code></dd><dt><a class="struct" href="struct.UQuote_.html" title="struct mcc_syntax::ast::symbols::UQuote_">UQuote_</a></dt><dd>Typed node <code>u'</code></dd></dl></section></div></main></body></html>