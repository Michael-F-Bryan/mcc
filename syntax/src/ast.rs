use codespan::ByteSpan;
use crate::node_id::NodeId;
use heapsize::HeapSizeOf;
use heapsize_derive::HeapSizeOf;
use std::any::Any;
use std::fmt::Debug;
use sum_type::*;

pub trait AstNode: Any + HeapSizeOf + Debug {
    fn span(&self) -> ByteSpan;
    fn node_id(&self) -> NodeId;
}

/// The result of parsing an entire source file.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct File {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub items: Vec<Item>,
}

impl File {
    pub(crate) fn new(items: Vec<Item>, span: ByteSpan) -> File {
        File {
            items,
            span,
            node_id: NodeId::placeholder(),
        }
    }
}

sum_type! {
    /// A top-level item.
    #[derive(Debug, Clone, PartialEq, HeapSizeOf)]
    pub enum Item {
        Function,
        u32,
    }
}

/// A function definition.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct Function {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub signature: FnDecl,
    pub body: Vec<Statement>,
}

impl Function {
    pub(crate) fn new(signature: FnDecl, body: Vec<Statement>, span: ByteSpan) -> Function {
        Function {
            span,
            signature,
            body,
            node_id: NodeId::placeholder(),
        }
    }

    pub fn name(&self) -> &str {
        &self.signature.name.name
    }
}

/// A function signature.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct FnDecl {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub return_value: Type,
    pub name: Ident,
    pub args: Vec<Argument>,
}

impl FnDecl {
    pub(crate) fn new(
        name: Ident,
        return_value: Type,
        args: Vec<Argument>,
        span: ByteSpan,
    ) -> FnDecl {
        FnDecl {
            name,
            return_value,
            args,
            span,
            node_id: NodeId::placeholder(),
        }
    }
}

/// An identifier.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct Ident {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub name: String,
}

impl Ident {
    pub(crate) fn new(name: &str, span: ByteSpan) -> Ident {
        Ident {
            span,
            name: name.into(),
            node_id: NodeId::placeholder(),
        }
    }
}

/// A type.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub enum Type {
    Ident(Ident),
}

impl From<Ident> for Type {
    fn from(other: Ident) -> Type {
        Type::Ident(other)
    }
}

/// A return statement.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct Return {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub value: Option<Expression>,
}

impl Return {
    pub(crate) fn bare(span: ByteSpan) -> Return {
        Return {
            span,
            node_id: NodeId::placeholder(),
            value: None,
        }
    }

    pub(crate) fn value(value: Expression, span: ByteSpan) -> Return {
        Return {
            span,
            node_id: NodeId::placeholder(),
            value: Some(value),
        }
    }
}

sum_type! {
    /// Any statement.
    #[derive(Debug, Clone, PartialEq, HeapSizeOf)]
    pub enum Statement {
        Return,
        /// Dummy variant so we can use the `sum_type!()` macro.
        u32,
    }
}

sum_type! {
    /// Any expression.
    #[derive(Debug, Clone, PartialEq, HeapSizeOf)]
    pub enum Expression {
        Literal,
        BinaryOp,
    }
}

#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct BinaryOp {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub left: Box<Expression>,
    pub right: Box<Expression>,
    pub kind: BinaryOperator,
}

#[derive(Debug, Copy, Clone, PartialEq, HeapSizeOf)]
pub enum BinaryOperator {
    Add,
    Subtract,
}

/// A single function argument.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct Argument {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub name: Option<Ident>,
    pub ty: Type,
}

/// A literal value.
#[derive(Debug, Clone, PartialEq, HeapSizeOf)]
pub struct Literal {
    pub span: ByteSpan,
    pub node_id: NodeId,
    pub kind: LiteralKind,
}

impl Literal {
    pub(crate) fn new(kind: LiteralKind, span: ByteSpan) -> Literal {
        Literal {
            kind,
            span,
            node_id: NodeId::placeholder(),
        }
    }
}

sum_type! {
    /// The particular kind of literal.
    #[derive(Debug, Clone, PartialEq, HeapSizeOf)]
    pub enum LiteralKind {
        Float(f64),
        Integer(i64),
        String(String),
    }
}

/// Apply the same operation to each variant in an enum.
///
/// # Examples
///
/// ```rust
/// # #[macro_use]
/// # extern crate syntax;
/// #[derive(Debug)]
/// pub enum Foo {
///     First(u32),
///     Second(String),
///     Third(Box<Foo>),
/// }
///
/// # fn main() {
/// let foo = Foo::Second("Hello World!".into());
///
/// defer!(Foo, foo; First, Second, Third => |item| println!("{:?}", item));
/// // prints "Hello World"
/// # }
/// ```
#[macro_export]
macro_rules! defer {
    ($type:ident, $this:expr; $( $variant:tt ),* => |$item:ident| $process_item:expr) => {{
        use self::$type::*;

        #[allow(unreachable_patterns)]
        match $this {
            $(
                $variant(ref $item) => $process_item,
            )*
            _ => unreachable!(),
        }
    }};
}

macro_rules! impl_ast_node {
    ($type:ty) => {
        impl $crate::ast::AstNode for $type {
            fn span(&self) -> ByteSpan {
                self.span
            }

            fn node_id(&self) -> NodeId {
                self.node_id
            }
        }
    };

    ($type:ident; $( $variant:ident ),+) => {
        impl $crate::ast::AstNode for $type {
            fn span(&self) -> ByteSpan {
                defer!($type, self; $($variant),+ => |item| item.span())
            }

            fn node_id(&self) -> NodeId {
                defer!($type, self; $($variant),+ => |item| item.node_id())
            }
        }
    };
}

impl_ast_node!(Argument);
impl_ast_node!(File);
impl_ast_node!(FnDecl);
impl_ast_node!(Function);
impl_ast_node!(Ident);
impl_ast_node!(Literal);
impl_ast_node!(Return);
impl_ast_node!(BinaryOp);
impl_ast_node!(Item; Function);
impl_ast_node!(Statement; Return);
impl_ast_node!(Expression; Literal, BinaryOp);
impl_ast_node!(Type; Ident);
