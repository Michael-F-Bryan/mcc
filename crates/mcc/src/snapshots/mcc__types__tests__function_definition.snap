---
source: crates/mcc/src/types.rs
description: void main();
expression: formatted
---
(translation_unit
  (declaration
    type: (primitive_type)
    declarator: (function_declarator
      declarator: (identifier)
      parameters: (parameter_list
        ("(")
        (")")))
    (";")))
